### PR TITLE
Recipe: Reinforcement learning for generative models using MindSpeed-mm as the backend (DanceGRPO)

### PR DESCRIPTION
**Pull request by ruanhao566**
This RFC proposes adding a comprehensive recipe to the verl-recipe repository that demonstrates how to integrate the verl reinforcement learning framework with [MindSpeed MM](https://gitcode.com/Ascend/MindSpeed-MM/blob/master/README.md) (Ascend Multi-modal Model Suite) to implement RLHF (Reinforcement Learning from Human Feedback) training for generative models on NPU (Ascend AI Processors). The recipe will feature the [DanceGRPO](https://dancegrpo.github.io/) algorithm, using [Wan2.2-5B](https://github.com/Wan-Video/Wan2.2) as the policy model and [HPSv3](https://github.com/MizzenAI/HPSv3) as the reward model, providing an end-to-end training workflow.

### CODE REVIEW

File: mm_dance_grpo/diffusion_worker.py [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+        # 重复推理当前文本
+        prompt = 'The video presents a serene winter landscape that remains consistent throughout its duration. It begins showcasing breathtaking scenery that includes snow-capped mountains, a mirror-like body of water reflecting the mountains and sky, snowflakes falling gently, and coniferous trees lining the shoreline. The scene is marked by a tranquil ambiance, with no noticeable changes as the video progresses. The mountains, water, snowflakes, and trees maintain their picturesque and motionless state, creating a continuous and peaceful winter wonderland. The overall composition of the landscape evokes a sense of tranquility and natural beauty, with the elements of the scene—light, shadow, and color—remaining unchanged, reinforcing the stillness and serenity of the winter setting.'
+        images_path = self.rollout.generate_test(prompt, step, save_path)
+        # 获取第一帧打分
+        image = video_first_frame_to_pil(images_path)
+        base, _ = os.path.splitext(images_path)
+        png_path = base + '.png'
+        image.save(png_path, format="PNG")
+        hps_score = self.reward_model.reward([png_path], [prompt])
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/diffusion_worker.py]: done

File: mm_dance_grpo/utils/rl_latent_dataset.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+# Copyright 2024 Bytedance Ltd. and/or its affiliates
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/utils/rl_latent_dataset.py:1]: Fixed.

File: mm_dance_grpo/run_verl_dance.sh [hunk 3]
```diff
@@ -0,0 +1,102 @@
+ray stop --force
+
+source /usr/local/Ascend/ascend-toolkit/s et_env.sh
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/run_verl_dance.sh]: Fixed.

File: mm_dance_grpo/patches/rl_dataset.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
+# Copyright 2024 Bytedance Ltd. and/or its affiliates
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/patches/rl_dataset.py:1]: Fixed.

File: mm_dance_grpo/main_dance.py [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+            if config.algorithm.use_kl_in_reward or config.actor_rollout_ref.actor.use_kl_loss:
+                role = Role.ActorRolloutRef
+            else:
+                role = Role.ActorRollout
+            self.role_worker_mapping[role] = ray.remote(actor_rollout_cls)
+            self.mapping[role] = "global_pool"
+            return actor_rollout_cls, ray_worker_group_cls
+
+        if config.actor_rollout_ref.actor.strategy in {"fsdp", "fsdp2"}:
+            from recipe.dance_grpo_mm.diffusion_workers import DiffusionActorRolloutWorker, MllmWorker
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/main_dance.py]: Fixed.

File: mm_dance_grpo/examples/wan2.2/5B/t2v/get_train_args.sh [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+OUTPUT_ARGS="
+    --log-interval 1 \
+    --save-interval 10000 \
+    --eval-interval 10000 \
+    --eval-iters 10 \
+    --load $LOAD_PATH \
+    --ckpt-format torch_dcp \
+"
+
+torchrun $DISTRIBUTED_ARGS pretrain_sora.py \
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/examples/wan2.2/5B/t2v/get_train_args.sh]: Fixed.

File: mm_dance_grpo/diffusion_worker.py [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+            "reward": hps_score.item(),
+            "images_path": images_path,
+            "prompt": prompt
+        }
+        file_path = f"{save_path}/reward_result.jsonl"
+        with open(file_path, "a", encoding="utf-8") as f:
+            import json
+            f.write(json.dumps(context, ensure_ascii=False) + "\n")
+
+    def seve_rollout_result(self, hps_score, prompt, images_path, global_steps):
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/diffusion_worker.py]: fixed

File: mm_dance_grpo/dataset/data_process.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        return process_fn
+
+    train_dataset = train_dataset.map(function=make_map_fn_to_zj2("train"), with_indices=True)
+    test_dataset = test_dataset.map(function=make_map_fn_to_zj2("test"), with_indices=True)
+
+    local_dir = args.local_dir
+    train_dataset.to_parquet(os.path.join(local_dir, "train.parquet"))
+    test_dataset.to_parquet(os.path.join(local_dir, "test.parquet"))
+    print(f"local_dir:{local_dir}")
+
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/dataset/data_process.py:71]: fixed

File: mm_dance_grpo/main_dance.py [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+            f.write(content)
+        print(f"Success {filepath} remove @autocast")
+        return True
+    return False
+
+
+if __name__ == "__main__":
+    current_dir = os.path.dirname(os.path.abspath(__file__))
+    file_to_modify = f"{current_dir}/../../mindspeed_mm/models/common/embeddings/pos_embeddings.py"
+    remove_autocast_from_file(file_to_modify)
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/main_dance.py:302]: fixed

File: mm_dance_grpo/dance_ray_trainer.py [hunk 10]
```diff
@@ -0,0 +1,649 @@
...
+        )
+        return gen_batch
+
+    def _create_dataloader(self, train_dataset, val_dataset, collate_fn, train_sampler: Optional[Sampler]):
+        """
+        Creates the train and validation dataloaders.
+        """
+        # TODO: we have to make sure the batch size is divisible by the dp size
+
+        from recipe.mm_dance_grpo.main_dance import create_rl_dataset, create_rl_sampler
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/dance_ray_trainer.py]: fixed

File: mm_dance_grpo/dance_ray_trainer.py [hunk 11]
```diff
@@ -0,0 +1,649 @@
...
+        # load critic
+        if self.use_critic:
+            self.critic_wg.load_checkpoint(
+                critic_path, del_local_after_load=self.config.trainer.del_local_ckpt_after_load
+            )
+
+        # load dataloader,
+        dataloader_local_path = os.path.join(global_step_folder, "data.pt")
+        if os.path.exists(dataloader_local_path):
+            dataloader_state_dict = torch.load(dataloader_local_path, weights_only=False)
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/dance_ray_trainer.py]: fixed

File: mm_dance_grpo/config/actor.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    Args:
+        strategy (str): Training strategy set to 'fsdp' for Fully Sharded Data Parallel.
+        grad_clip (float): Gradient clipping threshold.
+        ulysses_sequence_parallel_size (int): [DEPRECATED] Ulysses sequence parallel size for long sequences.
+        entropy_from_logits_with_chunking (bool): Whether to compute entropy from logits
+            with chunking for memory efficiency.
+        entropy_checkpointing (bool): Whether to use gradient checkpointing for entropy computation.
+        fsdp_config (dict[str, Any]): Configuration for FSDP settings.
+        use_remove_padding (bool): Whether to remove padding tokens in inputs during training
+    """
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/config/actor.py:36]: fixed

File: mm_dance_grpo/diffusion_workers.py [hunk 13]
```diff
@@ -0,0 +1,525 @@
...
+        self.config = config
+        self.role = role
+        self._is_actor = role in ["actor", "hybrid", "actor_rollout"]
+        self._is_rollout = role in ["rollout", "hybrid"]
+        # Init mindspeed args
+        import mindspeed.args_utils as args_utils
+        with open(f'{self.config.actor.model_args_path}/mindspeed_args.pkl', 'rb') as f:
+            args_utils._MINDSPEED_ARGS = pickle.load(f)
+        with open(f'{self.config.actor.model_args_path}/mm_args.pkl', 'rb') as f:
+            args = pickle.load(f)
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/diffusion_workers.py:113]: As mentioned in the readme file, the mindspeed_args.pkl and mm_args.pkl files need to be generated by users themselves, and the risk is controllable.

File: mm_dance_grpo/main_dance.py [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+            if config.algorithm.use_kl_in_reward or config.actor_rollout_ref.actor.use_kl_loss:
+                role = Role.ActorRolloutRef
+            else:
+                role = Role.ActorRollout
+            self.role_worker_mapping[role] = ray.remote(actor_rollout_cls)
+            self.mapping[role] = "global_pool"
+            return actor_rollout_cls, ray_worker_group_cls
+
+        if config.actor_rollout_ref.actor.strategy in {"fsdp", "fsdp2"}:
+            from recipe.mm_dance_grpo.diffusion_workers import DiffusionActorRolloutWorker, MllmWorker
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/main_dance.py]: fixed
